Add sample set displays & selection controls to editor

bdach opened the pull request:
Below 20 custom sample sets, they are shown as ternary buttons.

https://github.com/user-attachments/assets/be9c2959-168e-4edc-9eed-0264553be745

Above 20 custom sample sets, they are shown in a dropdown (yes there are actual cases of this as I've been informed by the NAT; one example being [this mapset](https://osu.ppy.sh/beatmapsets/1018061#osu/2197383), among other keysounded mapsets)

https://github.com/user-attachments/assets/dcbfbd7b-0221-43b7-bc40-f2a8637187c0

As a bonus, to make determining what the heck is actually changing when adjusting these controls, the full set of applicable sounds now plays on adding/removing additions, changing their banks, as well as changing the custom set (if any).

For now there are no user-facing controls to add the samples to the map yourself, you have to know how to name the files and edit-externally them in yourself. *For now.*

---

RFC. Further progress towards https://github.com/ppy/osu/issues/35370.

bdach committed:
Show sample suffix on timeline pieces

bdach committed:
Make bank selection controls dropdowns instead of text boxes

Longstanding complaint. Text boxes make no sense if anything that isn't
the three legacy values gets deleted on save anyway.

bdach committed:
Add sample set selection controls to sample popovers

- Below 20 custom sample sets, they are shown as ternary buttons.
- Above 20 custom sample sets, they are shown in a dropdown (yes there
  are actual cases of this as I've been informed by the NAT; one example
  being https://osu.ppy.sh/beatmapsets/1018061#osu/2197383)

As a bonus, to make determining what the heck is actually changing when
adjusting these controls, the full set of applicable sounds now plays on
adding/removing additions, changing their banks, as well as changing the
custom set (if any).

For now there are no user-facing controls to add the samples to the map
yourself, you have to know how to name the `.wav`s and edit-externally
them in yourself. *For now.*

bdach committed:
Inherit sample set info from previous object on placement

bdach committed:
Fix tests

bdach committed:
Fix code quality

God does `dotnet format` make the `..` spread syntax just absolutely
unusable with its stupid inspections.

minetoblend commented:
I gave this a little test drive and UI wise I really like it. 

Only thing I noticed is that I think the preview samples are getting played multiple times per button click.
When you click one of the sample buttons multiple times it sounds slightly different on the first click, likely as a result of this.

I recorded a video of it with all other ui sounds muted & log statements for each sample playback. I'm pretty sure only `SampleEditPopover.playDemoSample` actually produces a sound though, as `SampleSetTernaryButton.DemoSamples` appears to never be assigned.

https://github.com/user-attachments/assets/d336d394-0dd2-4181-b9d2-2372548f259e

bdach committed:
Remove stuff that was not even supposed to exist

bdach committed:
Fix playing sample multiple times

bdach commented:
> Only thing I noticed is that I think the preview samples are going off multiple times per button click. When you click one of the sample buttons multiple times it sounds slightly different on the first click, likely as a result of this.

Probably fixed in https://github.com/ppy/osu/pull/35430/commits/8755a016229af4126d3c568d771daeda368ac4f4?

> `SampleSetTernaryButton.DemoSamples` appears to never be assigned.

That shouldn't even exist, I forgot to remove it. Removed in https://github.com/ppy/osu/pull/35430/commits/901044d6902ddd13a4027cea0a943d9393adc6fb.

peppy committed:
Merge branch 'master' into custom-sample-set-selection

peppy committed:
Remove button sound from sample toggle buttons

The new ones added in this change don't play the UI sounds, likely
because they make it nigh impossible to actually hear what the gameplay
samples will sound like.

This removes the same sounds from the existing buttons to match.
